app/backend/src/couchers/servicers/api.py

```diff
@@ -657,6 +671,38 @@ def SendFriendRequest(self, request, context, session):
         ):
             context.abort(grpc.StatusCode.FAILED_PRECONDITION, errors.FRIENDS_ALREADY_OR_PENDING)
 
+        # Check if user has been sending friend requests excessively
+        count_friend_requests_last_24h = session.execute(
+            select(func.count())
+            .select_from(FriendRelationship)
+            .where(FriendRelationship.from_user_id == context.user_id)
+            .where(FriendRelationship.time_sent >= now() - timedelta(hours=24))
+        ).scalar_one()
+        if count_friend_requests_last_24h >= FRIEND_REQUEST_DAILY_BLOCKING_QUOTA - 1:
+            friend_requests = _get_user_friend_requests_in_past_time_interval(
+                session=session, user_id=context.user_id, interval=timedelta(hours=24)
+            )
+            send_friend_request_spam_report_email(
+                session=session,
+                user=user,
+                friend_requests=friend_requests,
+                threshold=FRIEND_REQUEST_DAILY_BLOCKING_QUOTA,
+                time_interval_str="24 hours",
+                user_is_blocked=True,
+            )
+            context.abort(grpc.StatusCode.RESOURCE_EXHAUSTED, errors.FRIEND_REQUEST_THRESHOLD)
+        if count_friend_requests_last_24h == FRIEND_REQUEST_DAILY_WARNING_QUOTA - 1:
```

Comment:
You are not guaranteed to catch this with equality. There can be race conditions, especially if someone is spamming. So you should update this somehow to use inequality.

Reply:
Is this something we have encountered with Couchers? But adjusted it to inequality, and storing RateLimitViolations in the DB